Update lightning imports to resolve issues related to change in pytorch_lightning naming

DouglasKrouth opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/DouglasKrouth/mlflow/pull/11372?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11372/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11372
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

<!-- Please fill in changes proposed in this PR. -->

### How is this PR tested?

- [ x ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [ x ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [ x ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ x ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ x ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

DouglasKrouth committed:
.gitignore update Signed-off-by: Douglas Krotuh <dkrouth@cibotechnologies.com>

Signed-off-by: DouglasKrouth <dkrouth@cibotechnologies.com>

DouglasKrouth committed:
updated imports in tests, formatting

Signed-off-by: DouglasKrouth <dkrouth@cibotechnologies.com>

DouglasKrouth committed:
Merge branch 'mlflow:master' into update_lightning_imports

DouglasKrouth committed:
fetch changes

Signed-off-by: DouglasKrouth <dkrouth@cibotechnologies.com>

serena-ruan commented on tests/pytorch/iris.py:
```diff
@@ -1,12 +1,25 @@
+# Handle import of either 'pytorch_lightning' or 'lightning'
+try:
+    import lightning.pytorch as pl
+except ModuleNotFoundError:
```
same here

serena-ruan commented on mlflow/pytorch/_lightning_autolog.py:
```diff
@@ -489,13 +522,18 @@ def patched_fit(original, self, *args, **kwargs):
             self.callbacks += [
                 __MLflowPLCallback(
-                    client, metrics_logger, run_id, log_models, log_every_n_epoch, log_every_n_step
+                    client,
```
Could we revert unrelated format change?

DouglasKrouth replied:
I unfortunately don't have control over this as it's enforced by Ruff on commit

harupy replied:
@DouglasKrouth We pin ruff to `0.2.1` in this repo to avoid unrelated changes like this. You can use safe-without-formatting.

serena-ruan commented on mlflow/ml-package-versions.yml:
```diff
@@ -49,6 +49,23 @@ pytorch:
     run: |
       pytest tests/pytorch/test_tensorboard_autolog.py
 
+lightning:
```
We don't need an extra flavor, if lightning and pytorch_lightning are the same package, we just need to replace pytorch_lightning with lightning. While for backwards compatibility test, we could inside `run` pip uninstall lightning and install pytorch_lightning to run the test again.

DouglasKrouth replied:
Removed extra flavor : c628e396173a4038402f7561b4b4cec3c0fa450e

serena-ruan commented on mlflow/pytorch/_lightning_autolog.py:
```diff
@@ -20,9 +20,32 @@
+
+        PYTORCH_LIGHTNING_LEGACY_NAMESPACE = True
+    except ModuleNotFoundError:
+        logging.error(
+            f"""'lightning' module not found, attempted use of alternative 'pytorch-lightning'\n
+            also failed. Please install 'lightning >={MIN_REQ_VERSION}'."""
+        )
```
We raise exception here, no need for extra logging

serena-ruan commented on mlflow/pytorch/_lightning_autolog.py:
```diff
@@ -20,9 +20,32 @@
+            also failed. Please install 'lightning >={MIN_REQ_VERSION}'."""
+        )
+        raise ModuleNotFoundError(
+            message=f"""Unable to import 'lightning' or 'pytorch-lightning'.
+            Please install lightning >= {MIN_REQ_VERSION} into your environment.\n
+            ('pip install lightning>={MIN_REQ_VERSION}')"""
+        )
```
```suggestion
            message="Unable to import 'lightning' or 'pytorch-lightning'."
            f"Please install lightning >= {MIN_REQ_VERSION} into your environment "
            f"with `pip install lightning>={MIN_REQ_VERSION}`"
        )
```

serena-ruan commented on mlflow/pytorch/_lightning_autolog.py:
```diff
@@ -38,10 +61,12 @@
+elif PYTORCH_LIGHTNING_LEGACY_NAMESPACE is True:
     from pytorch_lightning.utilities.model_summary import ModelSummary
+else:
+    from lightning.pytorch.utilities.model_summary import ModelSummary
```
Does lightning < 1.5.0 also imports from this namespace?
It would be clearer if we use `if PYTORCH_LIGHTNING_LEGACY_NAMESPACE, else` as the outer check and check version inside.

serena-ruan replied:
Does this comment apply `Does lightning < 1.5.0 also imports from this namespace?`?

serena-ruan commented on mlflow/pytorch/_lightning_autolog.py:
```diff
@@ -72,11 +99,18 @@ class __MLflowPLCallback(pl.Callback, metaclass=ExceptionSafeAbstractClass):
         if log_every_n_step and _pl_version < Version("1.1.0"):
             raise MlflowException(
-                "log_every_n_step is only supported for PyTorch-Lightning >= 1.1.0"
+                """log_every_n_step is only supported for Lightning >= 1.1.0\n
+                (and PyTorch-Lightning >= 1.1.0 backwards-compatible)"""
```
```suggestion
                "log_every_n_step is only supported for Lightning >= 1.1.0"
```

serena-ruan commented on .gitignore:
```diff
@@ -74,6 +74,8 @@ ENV/
 env.bak/
 venv.bak/
 .python-version
+bin/
+pyvenv.cfg
```
Why would there a bin/ folder here?

serena-ruan reviewed (commented):
Thanks for the PR! I left some comments :D

DouglasKrouth committed:
cleaned recipe

Signed-off-by: DouglasKrouth <dkrouth@cibotechnologies.com>

DouglasKrouth committed:
Update mlflow/pytorch/_lightning_autolog.py

Co-authored-by: Serena Ruan <82044803+serena-ruan@users.noreply.github.com>
Signed-off-by: Douglas <48629993+DouglasKrouth@users.noreply.github.com>

DouglasKrouth committed:
Update mlflow/pytorch/_lightning_autolog.py

Co-authored-by: Serena Ruan <82044803+serena-ruan@users.noreply.github.com>
Signed-off-by: Douglas <48629993+DouglasKrouth@users.noreply.github.com>

DouglasKrouth committed:
Reverted tests to remove try/catch clauses for py lightning backwards compat

Signed-off-by: DouglasKrouth <dkrouth@cibotechnologies.com>

DouglasKrouth committed:
fetch

Signed-off-by: DouglasKrouth <dkrouth@cibotechnologies.com>

DouglasKrouth committed:
Revert "Reverted tests to remove try/catch clauses for py lightning backwards compat"

This reverts commit c44e704d8b8b735a5ce1c8ddf38907b39c8ae90e.

Signed-off-by: DouglasKrouth <dkrouth@cibotechnologies.com>

DouglasKrouth committed:
Fixed changes to mlflow tests related to pl import

Signed-off-by: DouglasKrouth <dkrouth@cibotechnologies.com>

DouglasKrouth committed:
pr cleanup, changed cond check for pl legacy version in pl autolog

Signed-off-by: DouglasKrouth <dkrouth@cibotechnologies.com>

DouglasKrouth commented:
@serena-ruan went through comments, reverted changes to testing. Updated conditional check for pl legacy flag in _autolog file.

Had a couple issues with reverting the test changes which is why the commit history looks a bit wonky :laughing:

DouglasKrouth committed:
ruff formatting

Signed-off-by: DouglasKrouth <dkrouth@cibotechnologies.com>

serena-ruan commented on mlflow/pytorch/_lightning_autolog.py:
```diff
@@ -20,9 +20,28 @@
+    import lightning.pytorch as pl
+    from lightning.pytorch.utilities import rank_zero_only
+except ModuleNotFoundError:
+    logging.warning("'lightning' package not found, attempting to use 'pytorch_lightning instead.'")
+    # TODO : Using a nested try/catch isn't always nice, might be a better way to handle this?
```
Let's remove this

serena-ruan commented on mlflow/pytorch/_lightning_autolog.py:
```diff
@@ -20,9 +20,28 @@
+        import pytorch_lightning as pl
+        from pytorch_lightning.utilities import rank_zero_only
+
+        PYTORCH_LIGHTNING_LEGACY_NAMESPACE = True
```
Add a warning here instead
_logger.warning("'pytorch_lightning' will be deprecated in a future release, please use 'lightning' instead. `pip install lightning`)

serena-ruan commented on mlflow/pytorch/_lightning_autolog.py:
```diff
@@ -20,9 +20,28 @@
+        PYTORCH_LIGHTNING_LEGACY_NAMESPACE = True
+    except ModuleNotFoundError:
+        raise ModuleNotFoundError(
+            message="Unable to import 'lightning' or 'pytorch-lightning'."
```
```suggestion
            message="Unable to import 'lightning'. "
```

DouglasKrouth committed:
Update _lightning_autolog.py

Co-authored-by: Serena Ruan <82044803+serena-ruan@users.noreply.github.com>
Signed-off-by: Douglas <48629993+DouglasKrouth@users.noreply.github.com>

DouglasKrouth committed:
Update _lightning_autolog.py to change logger warning

Signed-off-by: Douglas <48629993+DouglasKrouth@users.noreply.github.com>

DouglasKrouth committed:
reverted change related to ruff auto formatting on __init__ signature

Signed-off-by: Douglas <48629993+DouglasKrouth@users.noreply.github.com>

serena-ruan commented:
@DouglasKrouth Thanks for the changes! Could you also update https://github.com/mlflow/mlflow/blob/79db4bc0ba7d147f168d5194ec0f91a5f7a8e442/mlflow/ml-package-versions.yml#L56-L61 to use lightning instead? We can trigger the workflow after this change :)

DouglasKrouth committed:
Updated lightning package version in yml spec

Signed-off-by: DouglasKrouth <dkrouth@cibotechnologies.com>

DouglasKrouth committed:
Merge branch 'update_lightning_imports' of github.com:DouglasKrouth/mlflow into update_lightning_imports

Signed-off-by: DouglasKrouth <dkrouth@cibotechnologies.com>

DouglasKrouth commented:
@serena-ruan updated the package spec here dc68ae83ada92219219cf4de5edceb78d59a57e4

DouglasKrouth commented:
@serena-ruan or @harupy Just double-checking if anything else is needed here?
